src/retrace/hooks/hooks.py

```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/python3
+
+import os
+import shlex
+from subprocess import PIPE, CalledProcessError, run, TimeoutExpired
+
+from retrace.retrace import log_info, log_error, log_debug
+from .config import HOOK_PATH, HOOK_TIMEOUT, hooks_config
+
+"""
+    Hooks description:
+    pre_start -- When self.start() is called
+    start -- When task type is determined and the main task starts
+    pre_prepare_debuginfo -- Before the preparation of debuginfo packages
+    post_prepare_debuginfo -- After the preparation of debuginfo packages
+    pre_prepare_mock -- Before the preparation of mock environment
+    post_prepare_mock -- After the preparation of mock environment
+    pre_retrace -- Before starting of the retracing itself
+    post_retrace -- After retracing is done
+    success -- After retracing success
+    fail -- After retracing fails
+    pre_remove_task -- Before removing task
+    post_remove_task -- After removing task
+    pre_clean_task -- Before cleaning task
+    post_clean_task -- After cleaning task
+"""
+
+
+def get_executables(path):
+    """ Scan `path` and return list of found executable scripts.
+    """
+    def _getname(entry):
+        return entry.name
+
+    script_list = []
+
+    if not os.path.isdir(path):
+        return script_list
+
+    with os.scandir(path) as dirls:
+        for entry in sorted(dirls, key=_getname):
+            if entry.is_file() and os.access(entry.path, os.X_OK):
+                script_list.append(entry)
+
+    return script_list
+
+
+class RetraceHook:
+
+    def __init__(self, task):
+        self.taskid = task.get_taskid()
+        self.task_results_dir = task.get_results_dir()
+
+    def _get_cmdline(self, hook, exc=None):
+        if exc:
+            cmdline = hooks_config.get(f"{hook}.{exc}.cmdline", None)
+
+        if not cmdline:
+            cmdline = hooks_config.get(f"{hook}.cmdline", None)
+
+        if cmdline:
+            cmdline = cmdline.format(hook_name=hook,
+                                     taskid=self.taskid,
+                                     task_results_dir=self.task_results_dir)
+
+        return cmdline
+
+    def _get_hookdir(self):
+        hooks_path = hooks_config.get("main.hookdir", HOOK_PATH)
+
+        return hooks_path
+
+    def _get_timeout(self, hook):
+        timeout = hooks_config.get("main.timeout", HOOK_TIMEOUT)
+
+        if f"{hook}.timeout" in hooks_config:
+            timeout = hooks_config.get(f"{hook}.timeout", timeout)
+
+        return int(timeout)
+
+    def run(self, hook):
+        """Called by the default hook implementations"""
+        hook_path = os.path.join(self._get_hookdir(), hook)
+        executables = get_executables(hook_path)
+        timeout = self._get_timeout(hook)
+
+        for exc in executables:
```

Comment:
When I originally did this loop I had a similar sequential loop.  But I thought it would be better to start all executable scripts in parallel and then wait for the exits.  You may want to consider something like this:
``    pids = []
    for f in os.listdir(h):
        if os.path.isfile(h + "/" + f) and os.access(h + "/" + f, os.X_OK):
            logging.info("Executing hook script %s/%s - max runtime %d seconds)" % (h, f, HOOK_RUNTIME_MAX))
            os.chdir(h)
            p = subprocess.Popen(["./" + f, args['task_id'], args['task_results_dir']],
                                 stdout=subprocess.PIPE, stderr=subprocess.PIPE)
            pids.append(p)
            os.chdir(cwd)

    logging.info("Waiting for all hooks to exit")
    count = HOOK_RUNTIME_MAX
    while len(pids) and count:
        for p in pids:
            if p.poll() is not None:
                logging.info("Hook pid %d exited with %d while running %s" %\
                             (p.pid, p.returncode, p.args))
                (out, err) = p.communicate()
                logging.info(out.decode("utf-8", "ignore"))
                logging.error(err.decode("utf-8", "ignore"))
                pids.remove(p)
        time.sleep(1)
        count -= 1

    if len(pids):
        for p in pids:
            p.kill()
            logging.warn("Hook pid %d got killed - %d second max runtime exceeded while running %s" %\
                         (p.pid, HOOK_RUNTIME_MAX, p.args))

``

Reply:
Done.